ports/nrf/common-hal/busio/I2C.c

```diff
@@ -29,67 +30,103 @@
 #include "py/mperrno.h"
 #include "py/runtime.h"
 
-#include "pins.h"
-#include "nrf.h"
+#include "nrfx_twim.h"
+#include "nrf_gpio.h"
+
+#define INST_NO 0
+
+static uint8_t twi_error_to_mp(const nrfx_err_t err) {
+    switch (err) {
+    case NRFX_ERROR_DRV_TWI_ERR_ANACK:
+        return MP_ENODEV;
+    case NRFX_ERROR_BUSY:
+        return MP_EBUSY;
+    case NRFX_ERROR_DRV_TWI_ERR_DNACK:
+    case NRFX_ERROR_INVALID_ADDR:
+        return MP_EIO;
+    default:
+        break;
+    }
+
+    return 0;
+}
 
-void common_hal_busio_i2c_construct(busio_i2c_obj_t *self, const mcu_pin_obj_t* scl, const mcu_pin_obj_t* sda, uint32_t frequency, uint32_t timeout ) {
-    if (scl->pin == sda->pin) {
+void common_hal_busio_i2c_construct(busio_i2c_obj_t *self, const mcu_pin_obj_t *scl, const mcu_pin_obj_t *sda, uint32_t frequency, uint32_t timeout) {
+    if (scl->pin == sda->pin)
         mp_raise_ValueError("Invalid pins");
-    }
 
-    NRF_GPIO->PIN_CNF[scl->pin] = ((uint32_t)GPIO_PIN_CNF_DIR_Input        << GPIO_PIN_CNF_DIR_Pos)
-                                    | ((uint32_t)GPIO_PIN_CNF_INPUT_Disconnect << GPIO_PIN_CNF_INPUT_Pos)
-                                    | ((uint32_t)GPIO_PIN_CNF_PULL_Disabled    << GPIO_PIN_CNF_PULL_Pos)
-                                    | ((uint32_t)GPIO_PIN_CNF_DRIVE_S0S1       << GPIO_PIN_CNF_DRIVE_Pos)
-                                    | ((uint32_t)GPIO_PIN_CNF_SENSE_Disabled   << GPIO_PIN_CNF_SENSE_Pos);
-
-    NRF_GPIO->PIN_CNF[sda->pin] = ((uint32_t)GPIO_PIN_CNF_DIR_Input        << GPIO_PIN_CNF_DIR_Pos)
-                                    | ((uint32_t)GPIO_PIN_CNF_INPUT_Disconnect << GPIO_PIN_CNF_INPUT_Pos)
-                                    | ((uint32_t)GPIO_PIN_CNF_PULL_Disabled    << GPIO_PIN_CNF_PULL_Pos)
-                                    | ((uint32_t)GPIO_PIN_CNF_DRIVE_S0S1       << GPIO_PIN_CNF_DRIVE_Pos)
-                                    | ((uint32_t)GPIO_PIN_CNF_SENSE_Disabled   << GPIO_PIN_CNF_SENSE_Pos);
-
-    // 1 for I2C, 0 for SPI
-    self->twi = NRF_TWIM1;
-
-    if ( frequency < 100000 ) {
-      self->twi->FREQUENCY = TWIM_FREQUENCY_FREQUENCY_K100;
-    }else if ( frequency < 250000 ) {
-      self->twi->FREQUENCY = TWIM_FREQUENCY_FREQUENCY_K250;
-    }else {
-      self->twi->FREQUENCY = TWIM_FREQUENCY_FREQUENCY_K400;
+    const nrfx_twim_t instance = NRFX_TWIM_INSTANCE(INST_NO);
+    self->twim = instance;
+
+    nrfx_twim_config_t config = NRFX_TWIM_DEFAULT_CONFIG;
+    config.scl = NRF_GPIO_PIN_MAP(scl->port, scl->pin);
+    config.sda = NRF_GPIO_PIN_MAP(sda->port, sda->pin);
+
+    // change freq. only if it's less than the default 400K
+    if (frequency < 100000) {
+        config.frequency = NRF_TWIM_FREQ_100K;
+    } else if (frequency < 250000) {
+      config.frequency = NRF_TWIM_FREQ_250K;
     }
 
-    self->twi->ENABLE = (TWIM_ENABLE_ENABLE_Enabled << TWIM_ENABLE_ENABLE_Pos);
+    nrfx_err_t err = nrfx_twim_init(&self->twim, &config, NULL, NULL);
```

Comment:
The config struct is interesting. Its very ASF3 style. They moved away from it because it encodes a bunch of data that's usually the default state of the peripheral once its software reset.

How does this impact code size?

Reply:
Actually, I don't think it's that much of an overhead, as the config structs are not that large, they only have the fields that are really needed and I'm not sure if we can trust the peripherals to always be in a known state, so probably best to init the bare minimum always. The benfit of using the drivers instead of the hal is that (in addition to being easier to use), they also have fixes/workarounds for known silicon bugs and and general error handling that we don't have to think about.

Reply:
You should measure it. :-)

I looked at the data sheet and there isn't a reset register like the SAMD peripherals have. So it makes sense to set everything.

In general I like this library and using it. I'm just giving you a heads up about our similar experiences with ASF3. Error checking is good but can add overhead if we know a given state but the compiler doesn't recognize. We have plenty of space now so starting with this is fine.